rayon-demo/src/life/mod.rs

```diff
@@ -0,0 +1,172 @@
+const USAGE: &'static str = "
+Usage: life bench [--size N] [--gens N]
+       life --help
+Conway's Game of Life.
+
+Commands:
+    bench           Run the benchmark in different modes and print the timings.
+Options:
+    --size N        Size of the game board (N x N) [default: 200]
+    --gens N        Simulate N generations [default: 100]
+    -h, --help      Show this message.
+";
+
+
+use rand::{thread_rng, Rng};
+use std::iter::repeat;
+use std::num::Wrapping;
+use time;
+
+use rayon::prelude::*;
+use docopt::Docopt;
+
+#[cfg(test)]
+mod bench;
+
+#[derive(RustcDecodable)]
+pub struct Args {
+    cmd_bench: bool,
+    flag_size: usize,
+    flag_gens: usize,
+}
+
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct Board {
+    board: Vec<bool>,
+    survive: Vec<usize>,
+    born: Vec<usize>,
+    rows: usize,
+    cols: usize
+}
+
+impl Board {
+    pub fn new(rows: usize, cols: usize) -> Board {
+        let born = vec![3];
+        let survive = vec![2, 3];
+
+        Board::new_with_custom_rules(rows, cols, born, survive)
+    }
+
+    fn new_with_custom_rules(rows: usize, cols: usize, born: Vec<usize>, survive: Vec<usize>) -> Board {
+        let new_board = repeat(false).take(rows * cols).collect();
+
+        Board { board  : new_board,
+                born   : born,
+                survive: survive,
+                rows   : rows,
+                cols   : cols }
+    }
+
+    fn len(&self) -> usize {
+        self.rows * self.cols
+    }
+
+    fn next_board(&self, new_board: Vec<bool>) -> Board {
+        assert!(new_board.len() == self.len());
+
+        Board { board  : new_board,
+                born   : self.born.clone(),
```

Comment:
Yeah, I even think they can be Rc, since they aren't touched by the parallel bit.